spiketoolkit/postprocessing/features.py

```diff
@@ -96,149 +115,202 @@ def compute_unit_template_features(waveforms,
+#                    timestamps[peak_idx]
+#     else:
+#         duration = timestamps[trough_idx:][np.where(waveform[trough_idx:] == np.max(waveform[trough_idx:]))[0][0]] - \
+#                    timestamps[trough_idx]
+#
+#     return duration * 1e3
 
-    return repolarization_slope * 1e-6
 
+def compute_trough_to_peak_duration(sorting,
+                                    unit_ids=None,
+                                    invert_template=False,
+                                    save_property_or_features=True,
+                                    return_through_and_peak_idx=False):
 
-def compute_unit_template_recovery_slopes(waveform, timestamps, window=20):
-    """
-    Spike recovery slope (after repolarization)
-    Inputs:
-    ------
-    waveform : numpy.ndarray (N samples)
-    timestamps : numpy.ndarray (N samples)
-    window : int
-        Window (in samples) for linear regression
-    Outputs:
-    --------
-    recovery_slope : slope of recovery period (V / s)
-    """
+    if isinstance(unit_ids, (int, np.integer)):
+        unit_ids = [unit_ids]
+    elif unit_ids is None:
+        unit_ids = sorting.get_unit_ids()
+    elif not isinstance(unit_ids, (list, np.ndarray)):
+        raise Exception("unit_ids should be an int or list")
 
-    max_point = np.argmax(np.abs(waveform))
+    assert 'template' in sorting.get_shared_unit_property_names(), 'add template to unit properties'
 
-    waveform = - waveform * (np.sign(waveform[max_point]))  # invert if we're using the peak
+    durations = []
+    peaks = []
+    throughs = []
 
-    peak_idx = np.argmax(waveform[max_point:]) + max_point
+    for uid in unit_ids:
+        templates = sorting.get_unit_property(uid, 'template')
+        unit_durations = []
+        unit_peaks = []
+        unit_throughs = []
 
-    recovery_slope = linregress(timestamps[peak_idx:peak_idx + window], waveform[peak_idx:peak_idx + window])[0]
+        for i, channel_template in enumerate(templates):
+            if invert_template:
+                channel_template = -channel_template
 
-    return recovery_slope * 1e-6
+            trough_idx = np.argmin(channel_template)
+            unit_throughs.append(trough_idx)
+            peak_idx = np.argmax(channel_template)
+            unit_peaks.append(peak_idx)
 
+            dur = (peak_idx - trough_idx) * (1 / sorting.get_sampling_frequency())
+            unit_durations.append(dur)  # exclude negative values?
 
-# ==========================================================
+        durations.append(unit_durations)
+        peaks.append(unit_peaks)
+        throughs.append(unit_throughs)
 
-# EXTRACTING 2D FEATURES
+    if save_property_or_features:
+        for i, uid in enumerate(unit_ids):
+            sorting.set_unit_property(uid, 'trough_to_peak_duration', durations[i])
+            sorting.set_unit_property(uid, 'peak_index_in_template', peaks[i])
```

Comment:
I would just save the `through_to_peak_duration` (the indexes are trivial and fast to recompute)

Reply:
I agree that's it's cheap to compute, but I was thinking maybe you might want to visualize in the future what your computations have done, at which point it's neater to not have the computation at a second place in your repository.

But what should be where, and how it's named, I'll leave up to you guys. For development I would like to keep it here, but once this module is finished we can easily strip these.

Reply:
👍